File: lab4/java/AVLMap.java
import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.ArrayList;
import java.util.function.Supplier;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * SimpleMap implemented as an AVL tree.
 * This inherits most functionality from BST, the only difference is 
 * that it rebalances the tree when adding new tree nodes.
 */
public class AVLMap<Key extends Comparable<Key>, Value> extends BSTMap<Key, Value> {

    AVLMap() {
        this(() -> null);
    }

    AVLMap(Supplier<Value> defaultValueSupplier) {
        this.root = null;
        this.defaultValueSupplier = defaultValueSupplier;
    }

    protected Node putHelper(Node node, Key key, Value value) {
        // Call `_put_helper` for normal BSTs, to insert the value into the tree,
        // and update the node to be the new parent.
        node = super.putHelper(node, key, value);

        // Calculate the balance factor for the node.
        int balanceFactor = this.heightHelper(node.left) - this.heightHelper(node.right);
        
        //---------- TASK 4a: Implement AVL rebalancing ---------------------//
        // Perform balance rotations if needed.
        // You do this by calling `leftRotate` and `rightRotate`.
        // Remember to return the new parent node if you balance the tree.

        if (balanceFactor > 1) {
            if (this.heightHelper(node.left.left) - this.heightHelper(node.left.right) < 0) {
                node.left = leftRotate(node.left);
            }
            node = rightRotate(node);
        } else if (balanceFactor < -1) {
            if (this.heightHelper(node.right.left) - this.heightHelper(node.right.right) > 0) {
                node.right = rightRotate(node.right);
            }
            node = leftRotate(node);
        }
        //---------- END TASK 4a -------------------------------------------//

        return node;
    }

    private Node rightRotate(Node parent) {
        // The left child will be the new parent.
        Node child = parent.left;
        // This must exist, otherwise we cannot rotate right.
        if (child == null) throw new NullPointerException("There must be a left child");

        //---------- TASK 4b: Implement right rotation ----------------------//
        // Don't forget to update the size and height for all nodes that need it.
        // You do this by calling the method `updateSizeAndHeight()`.
        parent.left = child.right;
        child.right = parent;
        parent.updateSizeAndHeight();
        child.updateSizeAndHeight();
        //---------- END TASK 4b --------------------------------------------//

        // Return the new parent.
        return child;
    }

    private Node leftRotate(Node parent) {
        // The left child will be the new parent.
        Node child = parent.right;
        // This must exist, otherwise we cannot rotate left.
        if (child == null) throw new NullPointerException("There must be a right child");

        //---------- TASK 4c: Implement left rotation -----------------------//
        // Don't forget to:
        //  1. Update the size and height for all nodes that need it.
        //     You do this by calling the method 'updateSizeAndHeight()`.
        //  2. Return the new parent node after rotation.
        parent.right = child.left;
        child.left = parent;
        parent.updateSizeAndHeight();
        child.updateSizeAndHeight();
        //---------- END TASK 4c --------------------------------------------//

        // Return the new parent.
        return child;
    }


    ///////////////////////////////////////////////////////////////////////////
    // Validation

    @Override
    public void validate() {
        super.validate();
        this.validateBalance(this.root);
    }

    private void validateBalance(Node node) {
        if (node == null) return;
        int balanceFactor = this.heightHelper(node.left) - this.heightHelper(node.right);
        if (! (-1 <= balanceFactor && balanceFactor <= 1))
            throw new IllegalArgumentException(String.format(
                "Node '%s:%s' not properly balanced: balance factor is %d, not -1, 0 or 1",
                node.key, node.value, balanceFactor
            ));
        this.validateBalance(node.left);
        this.validateBalance(node.right);
    }

    
    ///////////////////////////////////////////////////////////////////////////
    // Test the AVL tree: Change this code however you want.

    public static void main(String[] args) throws IOException {
        System.out.println("# A very simple tree.");
        AVLMap<Integer, String> tree = new AVLMap<>();
        tree.put(2, "dog"); tree.validate();
        tree.put(3, "sat"); tree.validate();
        tree.put(5, "the"); tree.validate();
        tree.put(6, "mat"); tree.validate();
        tree.put(4, "on");  tree.validate();
        tree.put(2, "cat"); tree.validate();
        tree.put(1, "the"); tree.validate();
        for (int k = 0; k < 8; k++) {
            System.out.format("%d --> %s\n", k, tree.get(k));
        }
        System.out.println(tree.show(5));
        System.out.println();

        System.out.println("# A tree where the values are mutable lists.");
        AVLMap<Integer, ArrayList<String>> tree2 = new AVLMap<>(() -> new ArrayList<>());
        tree2.get(1).add("the");
        tree2.get(3).add("sat");
        tree2.get(9).add("the");
        tree2.get(9).add("mat");
        tree2.get(3).add("on");
        tree2.get(1).add("cat");
        tree2.validate();
        for (int k : tree2) {
            System.out.format("%d --> %s\n", k, tree2.get(k));
        }
        System.out.println(tree2.show(5));
        if (tree2.size() != 3)
            throw new IllegalArgumentException(String.format("Wrong tree size: %d, but it should be 3", tree2.size()));
        System.out.println();

        
        // Wait with this until you're pretty certain that your code works.
        System.out.println("# A larger tree, testing performance.");
        AVLMap<Integer, Integer> tree3 = new AVLMap<>();
        List<Integer> numbers = IntStream.range(0, 5_000).boxed().collect(Collectors.toList());

        // Comment this to get an unbalanced tree:
        Collections.shuffle(numbers);

        for (int k : numbers) {
            tree3.put(k, k*k);  // Map a number to its square
        }
        tree3.validate();
        System.out.println(tree3.show(2));
        for (int k : numbers) {
            if (k*k != tree3.get(k))
                throw new IllegalArgumentException(String.format("%d --> %d", k, tree3.get(k)));
        }
        System.out.println();
       
    }

}

File: lab4/python/avl_map.py
from typing import Optional, List, Any

from bst_map import Node, BSTMap, Key, Value


class AVLMap(BSTMap[Key, Value]):
    """
    SimpleMap implemented as an AVL tree.
    This inherits most functionality from BST, the only difference is 
    that it rebalances the tree when adding new tree nodes.
    """

    def _putHelper(self, node: Optional[Node[Key, Value]], key: Key, value: Value) -> Node[Key, Value]:
        # Call `_putHelper` for normal BSTs, to insert the value into the tree,
        # and update the node to be the new parent.
        node = super()._putHelper(node, key, value)

        # Calculate the balance factor for the node.
        balanceFactor = self._heightHelper(node.left) - self._heightHelper(node.right)

        #---------- TASK 4a: Implement AVL rebalancing -------------------#
        # Perform balance rotations if needed.
        # You do this by calling `_left_rotate` and `_right_rotate`.
        # Remember to return the new parent node if you balance the tree.

        # TODO: Replace these lines with your solution!
        raise NotImplementedError()
        #---------- END TASK 4a ------------------------------------------#

        # If the tree doesn't need rebalancing, return the node as it is.
        return node


    def _rightRotate(self, parent: Node[Key, Value]) -> Node[Key, Value]:
        # The left child will be the new parent.
        child = parent.left
        # This must exist, otherwise we cannot rotate right.
        assert child, "There must be a left child"

        # Perform right rotation.
        #---------- TASK 4b: Implement right rotation --------------------#
        # Don't forget to update the size and height for all nodes that need it.
        # You do this by calling the method `updateSizeAndHeight()`.

        # TODO: Replace these lines with your solution! 
        raise NotImplementedError()
        #---------- END TASK 4b -------------------------------------------#

        # Return the new parent.
        return child


    def _leftRotate(self, parent: Node[Key, Value]) -> Node[Key, Value]:
        #---------- TASK 4c: Implement left rotation ---------------------#
        # Don't forget to:
        #  1. Update the size and height for all nodes that need it.
        #     You do this by calling the method 'update_size_and_height()'.
        #  2. Return the new parent node after rotation.

        # TODO: Replace these lines with your solution! 
        raise NotImplementedError()
        #---------- END TASK 4c ------------------------------------------#


    ###########################################################################
    # Validation

    def validate(self):
        super().validate()
        self._validateBalance(self.root)

    def _validateBalance(self, node: Optional[Node[Key, Value]]):
        if node is None: return
        balanceFactor = self._heightHelper(node.left) - self._heightHelper(node.right)
        if not (-1 <= balanceFactor <= 1):
            raise ValueError(f"Node '{node.key}:{node.value}' not properly balanced: "
                             f"balance factor is {balanceFactor}, not -1, 0 or 1")
        self._validateBalance(node.left)
        self._validateBalance(node.right)


###############################################################################
# Test the AVL tree: Change this code however you want.

def main():
    print("# A very simple tree.")
    tree: AVLMap[int, str] = AVLMap()
    tree.put(2, "dog"); tree.validate()
    tree.put(3, "sat"); tree.validate()
    tree.put(5, "the"); tree.validate()
    tree.put(6, "mat"); tree.validate()
    tree.put(4, "on");  tree.validate()
    tree.put(2, "cat"); tree.validate()
    tree.put(1, "the"); tree.validate()
    for k in range(1, 7):
        print(k, "-->", tree.get(k))
    print(tree.show(5))
    print()

    print("# A tree where the values are mutable lists.")
    tree2: AVLMap[int, List[Any]] = AVLMap(list)
    tree2.get(1).append("the")
    tree2.get(3).append("sat")
    tree2.get(9).append("the")
    tree2.get(9).append("mat")
    tree2.get(3).append("on")
    tree2.get(1).append("cat")
    tree2.validate()
    for k in tree2:
        print(k, "-->", tree2.get(k))        
    print(tree2.show(5))
    if (tree2.size() != 3):
        raise ValueError(f"Wrong tree size: {tree2.size()}, but it should be 3.")
    print()

    """
    # Wait with this until you're pretty certain that your code works.
    print("# A larger tree, testing performance.")
    tree3: AVLMap[int, int] = AVLMap()
    numbers = list(range(0, 2000))

    # Comment this to see an that the AVL tree doesn't become unbalanced:
    import random
    random.shuffle(numbers)

    for k in numbers:
        tree3.put(k, k*k)  # map a number to its square
    tree3.validate()
    print(tree3.show(2))
    for k in numbers:
        assert k*k == tree3.get(k)
    print()
    """


if __name__ == "__main__":
    main()
    

